include/fmt/core.h

```diff
@@ -1644,33 +1646,35 @@ template <typename Context> class basic_format_args {
     const format_arg* args_;
   };
 
-  bool is_packed() const { return (desc_ & detail::is_unpacked_bit) == 0; }
+  constexpr bool is_packed() const {
+    return (desc_ & detail::is_unpacked_bit) == 0;
+  }
   bool has_named_args() const {
     return (desc_ & detail::has_named_args_bit) != 0;
   }
 
-  detail::type type(int index) const {
+  FMT_CONSTEXPR detail::type type(int index) const {
     int shift = index * detail::packed_arg_bits;
     unsigned int mask = (1 << detail::packed_arg_bits) - 1;
     return static_cast<detail::type>((desc_ >> shift) & mask);
   }
 
-  basic_format_args(unsigned long long desc,
-                    const detail::value<Context>* values)
+  constexpr basic_format_args(unsigned long long desc,
+                              const detail::value<Context>* values)
       : desc_(desc), values_(values) {}
-  basic_format_args(unsigned long long desc, const format_arg* args)
+  constexpr basic_format_args(unsigned long long desc, const format_arg* args)
       : desc_(desc), args_(args) {}
 
  public:
-  basic_format_args() : desc_(0) {}
+  constexpr basic_format_args() : desc_(0), args_(nullptr) {}
 
   /**
    \rst
    Constructs a `basic_format_args` object from `~fmt::format_arg_store`.
    \endrst
    */
   template <typename... Args>
-  FMT_INLINE basic_format_args(const format_arg_store<Context, Args...>& store)
+  constexpr basic_format_args(const format_arg_store<Context, Args...>& store)
```

Comment:
Here and elsewhere: please keep `FMT_INLINE`.

Reply:
Yes, I missed that `FMT_INLINE` has an attribute inside, and this line can use `FMT_INLINE` **and** `constexpr`. But those lines where `FMT_INLINE` was replaced by `FMT_CONSTEXPR(20)` will have double `inline` specifier, which makes some compilers [unhappy](https://godbolt.org/z/8Exfo8) and some [very unhappy](https://godbolt.org/z/v8xq8e).

Reply:
So looks like one of these macros shouldn't set `inline` and that would be really strange if this macro is going to be `FMT_INLINE`.

I can give an example of [Hedley](https://github.com/nemequ/hedley), it does not fallback to `inline` in the case where `constexpr` is not available - https://github.com/nemequ/hedley/blob/5e50f6b735aeb4e09e3b2ad3d1717db3c0613bfa/hedley.h#L1269-L1276
and I think it's right because I can always mark a function as `inline` with or without `constexpr`, no matter that `constexpr` implies `inline`.

In case you agree with me here I can create a separate PR to remove `inline` from `FMT_CONSTEXPR(20)` macro. Otherwise I just don't know how to fix these problems without introducing `FMT_CONSTEXPR(20)_NO_INLINE` macro, which probably would be also bad naming because there is `FMT_NOINLINE` macro that expands to additional attribute...

Reply:
> In case you agree with me here I can create a separate PR to remove inline from FMT_CONSTEXPR(20) macro.

Sure.

Reply:
Done https://github.com/fmtlib/fmt/pull/2075